### ISSUE TITLE
Regression: Union of scalar and class

### ISSUE THREAD

**Issue by AndrolGenhald** [CuyZ/Valinor#375]
It used to be possible to map a type like `string|Foobar`, but with recent releases this now causes an error when trying to map `Foobar`:

```php
class Foobar
{
    public function __construct(public int $foo, public string $bar) {}
}

var_dump(
    (new CuyZ\Valinor\MapperBuilder())
        ->mapper()
        ->map("list<string|Foobar>", ["foo", ["foo" => 123, "bar" => "baz"], "bar"])
);
```

Expected output:

```
array(3) {
  [0]=>
  string(3) "foo"
  [1]=>
  object(Foobar)#96 (2) {
    ["foo"]=>
    int(123)
    ["bar"]=>
    string(3) "baz"
  }
  [2]=>
  string(3) "bar"
}
```

Actual output:

```
PHP Fatal error:  Uncaught CuyZ\Valinor\Mapper\TypeTreeMapperError: Could not map type `list<string|Foobar>`. An error occurred at path 1: Invalid value array{foo: 123, bar: 'baz'}. in Valinor/src/Mapper/TypeTreeMapper.php:29
```

`git bisect` points to 72cba320f582c7cda63865880a1cbf7ea292d2b1 as the cause.

**Comment by JanMikes** [comment]
Hi, im not sure if it is related to this issue or another bug:
```
/** @var string|list<SomeClass> */
```
Does not work for me. If strings are provided, it is okay, but as long as input is array, i get
```
CuyZ\Valinor\Mapper\TypeTreeMapperError: Could not map type `array<SomeClass>` with value array{0: array{…}, 1: array{…}, 2: array{…}, 3: array{…}, 4: array{…}, 5: array{…}, …}. A total of 19 errors were encountered.
```

**Comment by JanMikes** [comment]
Tried your code in your PR and it does not fix my issue, i will open new one.

**Comment by romm** [comment]
Should be fixed by #376.
